pkg/component/autoscaling/gardenercustommetrics/component.go

```diff
@@ -0,0 +1,215 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/component/autoscaling/gardenercustommetrics/kubeobjects"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+// ComponentName is the component name.
+const ComponentName = componentBaseName
+
+// GardenerCustomMetrics manages an instance of the gardener-custom-metrics component (aka GCMx). The component is
+// deployed on a seed, scrapes the metrics from all shoot kube-apiserver pods, and provides custom metrics by
+// registering as APIService at the custom metrics extension point of the seed kube-apiserver.
+// For information about individual fields, see the NewGardenerCustomMetrics function.
+type GardenerCustomMetrics struct {
+	namespaceName      string
+	containerImageName string
+	kubernetesVersion  *semver.Version
+
+	seedClient     client.Client
+	secretsManager secretsmanager.Interface
+}
+
+// NewGardenerCustomMetrics creates a new GardenerCustomMetrics instance tied to a specific server connection.
+//
+// namespace is where the server-side artefacts (e.g. pods) will be deployed (usually the 'garden' namespace).
+// containerImageName points to the binary for the gardener-custom-metrics pods. The exact version to be used, is
+// determined by contextual configuration, e.g. image vector overrides.
+// If enabled is true, this instance strives to bring the component to an installed, working state. If enabled is
+// false, this instance strives to uninstall the component.
+// seedClient represents the connection to the seed API server.
+// secretsManager is used to interact with secrets on the seed.
+func NewGardenerCustomMetrics(
+	namespace string,
+	containerImageName string,
+	kubernetesVersion *semver.Version,
+	seedClient client.Client,
+	secretsManager secretsmanager.Interface) *GardenerCustomMetrics {
+	return &GardenerCustomMetrics{
+		namespaceName:      namespace,
+		containerImageName: containerImageName,
+		kubernetesVersion:  kubernetesVersion,
+		seedClient:         seedClient,
+		secretsManager:     secretsManager,
+	}
+}
+
+// Deploy implements [component.Deployer.Deploy]()
+func (gcmx *GardenerCustomMetrics) Deploy(ctx context.Context) error {
+	baseErrorMessage :=
+		fmt.Sprintf(
+			"An error occurred while deploying gardener-custom-metrics component in namespace '%s' of the seed server",
+			gcmx.namespaceName)
+
+	serverCertificateSecret, err := gcmx.deployServerCertificate(ctx)
+	if err != nil {
+		return fmt.Errorf(baseErrorMessage+
+			" - failed to deploy the gardener-custom-metrics server TLS certificate to the seed server. "+
+			"The error message reported by the underlying operation follows: %w",
+			err)
```

Comment:
I understand that there is a mandatory consistency requirement to stick to the error message approaches used by the existing code. I have therefore refactored error handing to match those. However, since I believe that the approach I proposed has important advantages, below I will list my concerns:  

I intentionally avoided following some of the patterns I've seen in existing code. From the behavior perspective, I'd like to avoid the frustratingly common case where a perfectly good error message get detached from context to the point where it is nearly useless. E.g, from GRM I get:
>http: TLS handshake error from 55.66.77.88:59999: EOF

Certainly, the code must know what service it was trying to reach at "55.66.77.88:59999". If that information is made available to the operator, it would speed up troubleshooting. Unfortunately is looks like the code passed the error along, without supplementing it with context information. I would like to avoid that.

Another arbitrary example, this one from the code perspective: [ReadManagedSeedAPIServer()](https://github.com/gardener/gardener/blob/1b160a3c68b9c5b59ed75f171c77397e3306efc4/pkg/apis/core/v1beta1/helper/helper.go#L295) and its helpers up the call stack, [parseManagedSeedAPIServer()](https://github.com/gardener/gardener/blob/1b160a3c68b9c5b59ed75f171c77397e3306efc4/pkg/apis/core/v1beta1/helper/helper.go#L179) and parseManagedSeedAPIServerAutoscaler():
Multiple distinct configuration settings get parsed via strconv.ParseInt(). Errors get propagated without adding contextual information. So the operator ends up with a componsite configuration object rejected with a generic integer parsing fault. Upon error, the failing code knows exactly which of all the setting values is the offending one, but does not pass that info to the user. I think this wastes far more time in operations, than it saves in development.

The other thing I strive for with the pattern I used is: make the error understandable by the non-expert. E.g, look through the eyes of the unexperienced at a message like "observed generation outdated", as opposed to somehting as descriptive as:
```
An error occurred while reconciling the gardener-custom-metrics component in namespace 'garden' of the seed server - failed to deploy the prometheus metrics adapter server TLS certificate to the seed server.
The error message reported by the underlying operation follows:
  An error occurred while deploying server TLS certificate for the prometheus metrics adapter - the CA certificate, which is required to sign said server certificate, is missing. The CA certificate was expected in the 'ca-seed' secret, but that secret was not found
```

I think that for the non-expert, the proposed pattern could offer a great improvement in troubleshooting efficiency, over some existing messages such as:
'observed generation outdated'
'object's "gardener.cloud timestamp" annotation is not "2024-04-17T10:34:45" but "2024-04-17T10:31:34"'
'worker status machineDeployments has not been updated'

Overall, my reasoning is:
- Troubleshooting makes for a substantial contribution to TCO.
- Some operators are new to Gardener. Even veteran operators constantly face cases where some error message does not provide sufficient information to understand the root cause, and are forced to do costly research or even reach out to component maintainers.
- Errors do bubble up to the shoot dashboard - the end user is also exposed to error messages.
- The UX and cost of resolving faults are an important factor in both objective and perceived service quality.
- When troubleshooting a failure mode they haven't seen before, the user (even operator) practically does not mind a longer error message that would take two extra minutes to read. What they do mind, is a an error message which lacks pertinent information and forces upon them hours of digging into logs and code.
- Similarly, when faced with an error, the user does not care if the error message format is consistent with other error messages from the same application.
- End users are not necessarily familiar with golang's "aaa: bbb: ccc" error format. Natural language and standard capitalisation provides better utility to them. An error message is UI.
- The suggested error handling pattern automatically provides relevant expertise (the algorithmic causaility chain) precisely where it is needed, and in terms understandable to the user.
- I therefore proposed a pattern which, albeit verbose and currently lacking a few utilities necesasry to streamline it, if adopted would result in better Gardener operability (specifically troubleshooting).

**Anyway, in the interest of moving forward, I have refactored the error messages in the format requested by the review. The change will be available with the next push.**